ENT-5866: Fixed set_line_based() for case when edit_defaults.empty_before_use is true

nickanderson opened the pull request:
Merge together:

* https://github.com/cfengine/core/pull/4993
* https://github.com/cfengine/documentation/pull/2779

* new variable `edit.empty_before_use` in docs (http://buildcache.cfengine.com/packages/build-documentation-pr/jenkins-fast-build-and-deploy-docs-master-62/output/_site/reference-special-variables-edit.html#edit-empty_before_use)
** Cross referenced from `edit_defaults.empty_file_before_editing` (http://buildcache.cfengine.com/packages/build-documentation-pr/jenkins-fast-build-and-deploy-docs-master-62/output/_site/reference-promise-types-files.html#empty_file_before_editing)


http://buildcache.cfengine.com/packages/build-documentation-pr/jenkins-fast-build-and-deploy-docs-master-62/output/_site/reference-promise-types-files.html#empty_file_before_editing

The implementation now avoids looking in the original file when the files
promise instructed to disregard the files starting content (`edit_defaults => empty`) on cfengine versions where this is knowable from within edit_line bundles.

nickanderson commented:
@cf-bottom jenkins, please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=8311)](https://ci.cfengine.com//job/pr-pipeline/8311/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8311/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8311/

nickanderson commented:
Oh, it's failing travis I think because it's using a package from nightly, not with the changes from core.

nickanderson commented:
@cf-bottom jenkins, again please.

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=8312)](https://ci.cfengine.com//job/pr-pipeline/8312/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8312/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8312/

vpodzime reviewed (commented):
Looks promising to me. But aren't the macros not supported in 3.15.0 an issue because the policy won't even be syntactically correct there?

olehermanse reviewed (changes requested):
Please squash commits and test it, both with new versions and compatibility with 3.15.

craigcomstock commented:
@cf-bottom build please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=8314)](https://ci.cfengine.com//job/pr-pipeline/8314/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8314/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8314/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=8315)](https://ci.cfengine.com//job/pr-pipeline/8315/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8315/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8315/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=8316)](https://ci.cfengine.com//job/pr-pipeline/8316/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8316/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8316/

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=build-and-deploy-docs-master&build=30)](https://ci.cfengine.com//job/build-and-deploy-docs-master/30/)

**Jenkins:** https://ci.cfengine.com/job/build-and-deploy-docs-master/30/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-build-and-deploy-docs-master-30/

**Documentation:** http://buildcache.cfengine.com/packages/build-documentation-pr/jenkins-build-and-deploy-docs-master-30/output/_site/

craigcomstock reviewed (commented):
looks good, trying a build now.

nickanderson commented:
> Please squash commits and test it, both with new versions and compatibility with 3.15.

-   [X] Commits squashed
-   Test results
    -   Behavior on master (3.21.0)
        -   `set_line_based` does properly deals with the combination of `edit_defaults => empty` with a pre-existing file.
    -   Behavior on 3.15.0
        -   Policy works as it did prior to change.
            -   `set_line_based` does **not** work correctly when paired with `edit_defaults => empty`
                -   3.15.0 doesn't have any way to know that `edit_defaults => empty` is in use, thus no way to tweak the implementation to make it smart about differing the behavior.
                -   Prior version of fix resulted in syntax errors for the use of the `before_version` macro which did not exist in 3.15.0, removed it's use and adjusted class guard to handle backwards compatibility.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=build-and-deploy-docs-master&build=31)](https://ci.cfengine.com//job/build-and-deploy-docs-master/31/)

**Jenkins:** https://ci.cfengine.com/job/build-and-deploy-docs-master/31/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-build-and-deploy-docs-master-31/

**Documentation:** http://buildcache.cfengine.com/packages/build-documentation-pr/jenkins-build-and-deploy-docs-master-31/output/_site/

vpodzime commented on lib/files.cf:
```diff
@@ -768,10 +768,25 @@ bundle edit_line set_line_based(v, sep, bp, kp, cp)
+      unless => strcmp( "true", $(edit.empty_before_use) );
+@endif
+
+    (cfengine_3_15|cfengine_3_16|cfengine_3_17|cfengine_3_18|cfengine_19|cfengine_3_20)::
```
`cfengine_3_19`

nickanderson replied:
doh!

vpodzime reviewed (commented):
We already have an acceptance test for this added in https://github.com/cfengine/core/commit/a8453b1517ffa9be5f72f54d0fd8d988ddd65310 But I'd actually remove that one because `set_line_based()` tests don't really belong to the *core* repo. Looks good to me otherwise.

nickanderson commented:
> We already have an acceptance test for this added in [cfengine/core@a8453b1](https://github.com/cfengine/core/commit/a8453b1517ffa9be5f72f54d0fd8d988ddd65310) But I'd actually remove that one because `set_line_based()` tests don't really belong to the _core_ repo. Looks good to me otherwise.

OK, we can do that on your PR in core. lms if i can push to that branch.

nickanderson commented:
> > We already have an acceptance test for this added in [cfengine/core@a8453b1](https://github.com/cfengine/core/commit/a8453b1517ffa9be5f72f54d0fd8d988ddd65310) But I'd actually remove that one because `set_line_based()` tests don't really belong to the _core_ repo. Looks good to me otherwise.
> 
> OK, we can do that on your PR in core. lms if i can push to that branch.

I pulled it in your core PR: https://github.com/cfengine/core/pull/4993/commits/ebe7fefe0a2d6233bbd3aafd7ec4565461e097eb

vpodzime reviewed (approved):
ACK
